crates/driver/src/domain/competition/auction.rs

```diff
@@ -23,26 +34,52 @@ pub struct Auction {
 }
 
 impl Auction {
-    pub fn new(
+    pub async fn new(
         id: Option<Id>,
-        mut orders: Vec<competition::Order>,
+        orders: Vec<competition::Order>,
         tokens: impl Iterator<Item = Token>,
-        gas_price: eth::GasPrice,
         deadline: Deadline,
-        weth: eth::WethAddress,
-    ) -> Result<Self, InvalidTokens> {
+        eth: &Ethereum,
+    ) -> Result<Self, Error> {
         let tokens = Tokens(tokens.map(|token| (token.address, token)).collect());
 
         // Ensure that tokens are included for each order.
+        let weth = eth.contracts().weth_address();
         if !orders.iter().all(|order| {
             tokens.0.contains_key(&order.buy.token.wrap(weth))
                 && tokens.0.contains_key(&order.sell.token.wrap(weth))
         }) {
-            return Err(InvalidTokens);
+            return Err(Error::InvalidTokens);
         }
 
-        // Sort orders such that most likely to be fulfilled come first.
-        orders.sort_by_key(|order| {
+        Ok(Self {
+            id,
+            orders,
+            tokens,
+            gas_price: eth.gas_price().await?,
+            deadline,
+        })
+    }
+
+    /// [`None`] if this auction applies to a quote. See
+    /// [`crate::domain::quote`].
+    pub fn id(&self) -> Option<Id> {
+        self.id
+    }
+
+    /// The orders for the auction.
+    pub fn orders(&self) -> &[competition::Order] {
+        &self.orders
+    }
+
+    // Prioritize the orders such that those which are more likely to be fulfilled
+    // come before less likely orders. Filter out orders which the trader doesn't
+    // have enough balance to pay for.
+    //
+    // Prioritization is skipped during quoting. It's only used during competition.
+    pub async fn prioritize(mut self, eth: &Ethereum) -> Self {
```

Comment:
This comment makes me think that we should have 2 auction types:
1. that represents the `Auction` we get for the protocol to solve
2. the `Augmented` auction that is the result of:
  * fetching token data
  * fetching liquidity
  * prioritization and filtering
  
If we do have to separate types, then it becomes clear at a type level which "methods" are applied on an `Auction` for solving and for quoting:
* `async Auction -> auction::Augmented` transformation does order prioritization.
* `async quote::Order -> auction::Augmented` transformation does **not** do order prioritization.

**Not a suggestion for this PR**, though. Just a general suggestion to clean up (what is IMO) a code smell that is starting to appear as we add some more logic to these domain types.